### PR TITLE
Add expressions is_empty(geom)  is_empty_or_null(geom)

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

Following my work on empty geometries, here is a new expression to catch them.

image

sponsored by QWAT cc @kandre @haubourg 
## Checklist

- [X] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [X] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

**Comment by m-kuhn** [comment]
Nice!
What's the preferred way to test that a geometry is neither empty nor null? I think that should be listed in the docs.

**Comment by haubourg** [comment]
>  What's the preferred way to test that a geometry is neither empty nor null? I think that should be listed in the docs.

@m-kuhn well, the SQL way is `geom is Null or st_IsEmpty(geom)`, I suppose we should behave the same. 
However a all in one expression catching both NULL or empty would be nice to avoid this nasty little tricky empty notion nobody knows about. I have no clue how we could name it.

**Comment by roya0045** [comment]
> @m-kuhn well, the SQL way is `geom is Null or st_IsEmpty(geom)`, I suppose we should behave the same.
> However a all in one expression catching both NULL or empty would be nice to avoid this nasty little tricky empty notion nobody knows about. I have no clue how we could name it.

But isn't a Null geometry or an empty geometry the same thing to some extent. Since geometries are defined by their position an empty container or no container should be the same to any user. 

I don't think an empty geometry gives any advantage over a Null geometry as a user, please let me know if I'm wrong.

**Comment by m-kuhn** [comment]
It's an SQL thing. Just like `'' IS NOT NULL`. We shouldn't change it at this level, but make it easy for user to not get it wrong.

 - By documenting it clearly
 - And even better by adding a shorthand like `is_void(geometry)` (counter suggestions very welcome!).

File: resources/function_help/json/is_empty [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "is_empty",
+  "type": "function",
+  "description": "Returns true if a geometry is empty (without coordinates) or false if a is empty."
```
Comment (nyalldawson) [inline resources/function_help/json/is_empty]:
```suggestion
  "description": "Returns true if a geometry is empty (without coordinates) or false if the geometry is empty."
```

**Comment by nyalldawson** [comment]
>, the SQL way is geom is Null or st_IsEmpty(geom), I suppose we should behave the same.

+1 to "geom is null"

**Comment by m-kuhn** [comment]
> > , the SQL way is geom is Null or st_IsEmpty(geom), I suppose we should behave the same.
> 
> +1 to "geom is null"

What's that +1 for exactly? (or what's the counter proposal you are dismissing?)

**Comment by nyalldawson** [comment]
> What's that +1 for exactly? (or what's the counter proposal you are dismissing?)

It was in contrast to st_isempty(geom) as a null test. A geometry could be non-null but empty...

**Comment by lbartoletti** [comment]
> > What's that +1 for exactly? (or what's the counter proposal you are dismissing?)
> 
> It was in contrast to st_isempty(geom) as a null test. A geometry could be non-null but empty...

I will add another expression is_null(geom)

**Comment by nyalldawson** [comment]
Why not just plugin into the existing "IS NOT NULL" support?

**Comment by lbartoletti** [comment]
> Why not just plugin into the existing "IS NOT NULL" support?

I hadn't thought about that, I usually go through functions it seems clearer to me :)

**Comment by m-kuhn** [comment]
Any feedback on the shorthand for null-or-empty? I think providing and recommending this could act as a help to avoid common pitfalls (for the 99% of the cases where people don't care about the difference).

**Comment by lbartoletti** [comment]
> Any feedback on the shorthand for null-or-empty? I think providing and recommending this could act as a help to avoid common pitfalls (for the 99% of the cases where people don't care about the difference).

Oh sure, I can add a function grouping the is empty or null in a special function.

**Comment by lbartoletti** [comment]
> > Why not just plugin into the existing "IS NOT NULL" support?
> 
> I hadn't thought about that, I usually go through functions it seems clearer to me :)

@nyalldawson In fact, no needs to add an expression "$geometry IS NULL" works

**Comment by lbartoletti** [comment]
@m-kuhn FYI is_empty(NULL) (or null geometry) returns NULL too

**Comment by m-kuhn** [comment]
@lbartoletti sorry, I can't follow your latest statement and it's implications. I.e. what does (or null geometry) mean, and what does the "too" refer to?

**Comment by lbartoletti** [comment]
> @lbartoletti sorry, I can't follow your latest statement and it's implications. I.e. what does (or null geometry) mean, and what does the "too" refer to?

Sorry, I didn't see your message and I edited mine in the meantime.

**Comment by m-kuhn** [comment]
Thanks. This sounds reasonable as test result :+1: 
I still wonder what the "too" means in the message. And if this was meant as a response to the shorthand proposal.

File: resources/function_help/json/is_empty [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "is_empty",
+  "type": "function",
+  "description": "Returns true if a geometry is empty (without coordinates) or false if the geometry is empty.",
```
Comment (DelazJ) [inline resources/function_help/json/is_empty]: Isn't there an issue in this sentence? same condition returns either true or false?

Comment (lbartoletti) [reply resources/function_help/json/is_empty]: Done. Thanks...

File: resources/function_help/json/is_empty [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "is_empty",
+  "type": "function",
+  "description": "Returns true if a geometry is empty (without coordinates), false if the geometry is not empty and NULL if there is no geometry.",
```
Comment (m-kuhn) [inline resources/function_help/json/is_empty]:
```suggestion
  "description": "Returns true if a geometry is empty (without coordinates), false if the geometry is not empty and NULL if there is no geometry. See also `is_empty_or_null`.",
```

**Review by m-kuhn — APPROVED** [review]
Thanks a lot. One minor suggestion to bring people to the right path.

**Comment by DelazJ** [comment]
just to add my usual comment: `[feature] is missing` :smile:

**Comment by m-kuhn** [comment]
Sorry @DelazJ , I missed your comment.
I added a label.
@lbartoletti best to add it directly to the changelog already.

**Comment by DelazJ** [comment]
> Sorry @DelazJ , I missed your comment.
I added a label.

No problem. I don't think it works afterwards. Not yet. This was your grant proposal, right? Meanwhile, I think we need to manually create the report in docs repo
>@lbartoletti best to add it directly to the changelog already.

Do we have an automated changelog generated?

**Comment by m-kuhn** [comment]
> No problem. I don't think it works afterwards. Not yet. This was your grant proposal, right? Meanwhile, I think we need to manually create the report in docs repo

Yep, @3nids is on it. And it should also work for previous ones, one just needs to relabel them (so the important point is to be able to find them again)

> Do we have an automated changelog generated?

Semi automated by @timlinux at some point I think

**Pull request merged**